### PR TITLE
Fix early init of SCTLR for few ARMv7 platform

### PR DESCRIPTION
**Pull request by etienne-lms**
2 changes for the same P-R as intended to hit 2.5.0 and cross dependent.

* core: fix Cortex-A9 early init of SCTLR
    
    Prior this change, plat_cpu_reset_early() implementation discarded
    the generic effect of CFG_SCTLR_ALIGNMENT_CHECK.
    
    To overcome the issue and since these platforms only enable or not
    the round robin replacement strategy for the several caches (icache,
    dcache, btac, tlb); this change introduces a new configuration
    directive: CFG_SCTLR_ROUNDROBIN_REPLACEMENT. Now plat_cpu_reset_early()
    no more initializes the SCTLR.
    
    Note that the plat_cpu_reset_early() api is implemented only for
    supported Cortex-A9 based platforms.
    
    This change updates platforms mx6-(qsabrelite|qsabresd|dlsabresd),
    ls-ls1021atwr, stm-(cannes2|b2260) and zynq-zc702.
    
* core: allow enable of program flow prediction from early inits
    
    Introduce CFG_EARLY_BRANCH_PREDICTION to allow platform or external
    configuration the enable core program flow prediction from OP-TEE
    early inits.
    
    Early branch prediction may be unsafe against uncontrolled memory
    prefetches that could hit some hard memory access control firewalls.
    
    CFG_EARLY_BRANCH_PREDICTION is supported only for ARMv7 cores. ARMv8
    core automatically enable program flow prediction at MMU initialization
    and hence do not need to manually enable it from the SW.
    
    Some cores require btac flushing before enabling the program flow
    prediction. Flush btac unconditionally when enable the prediction
    block as it should not harm if not required: we're at early boot.
    Ref: ARMv7RM " ID_MMFR1, Memory Model Feature Register 1, PMSA".

### CODE REVIEW

**Comment by jforissier** [comment]
I'll let the i.MX maintainers comment further on this, but it looks good to me at least. A couple of minor issues/questions:
- The CFG names are quite long. This is often inconvenient. How about using the names ot the bits in the RM, that is: `s/CFG_SCTLR_ROUNDROBIN_REPLACEMENT/CFG_SCTLR_RR/` and `s/CFG_EARLY_BRANCH_PREDICTION/CFG_SCTLR_Z/`? In this case we should also rename `s/CFG_SCTLR_ALIGNMENT_CHECK/CFG_SCTLR_A` for consistency in a separate patch. I'm using the shorter names in the next bullet ;)
- When `CFG_SCTLR_RR` and `CFG_SCTLR_Z` are `y`, the bits are set ; but they are not cleared when the config is `n` (contrary to what happens with `CFG_SCTLR_A`). What is the correct way? The previous bootloader stage may have set any configuration I suppose, so one should not assume the bits are cleared when entering OP-TEE.

**Comment by etienne-lms** [comment]
I think `CFG_SCTLR_ROUNDROBIN_REPLACEMENT` pollutes the cfg. let the early_plat_init `orr` the bit if required. And don't add `CFG_SCTLR_RR`. No ?

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 1]
```diff
@@ -198,9 +198,18 @@ UNWIND(	.cantunwind)
...
 	orr	r0, r0, #(SCTLR_WXN | SCTLR_UWXN)
 #endif
+#if defined(CFG_SCTLR_ROUNDROBIN_REPLACEMENT)
+	orr	r0, r0, #SCTLR_RR
+#endif
 	write_sctlr r0
 	isb
 
+#if defined(CFG_EARLY_BRANCH_PREDICTION)
+	write_bpiall
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: This should be followed by
```
dsb
isb
```

**Comment by jenswi-linaro** [comment]
@etienne-lms I think it makes sense to have the round robin config in the generic code too.
As long as there's good prefixes to the config variables there's minimal pollution in the config system.

File: core/arch/arm/kernel/generic_entry_a32.S (hunk 1 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: these sync are not used when flushing btac in other places (cache, tlb).
Should it be fixed also ?

Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: I think it should be added to `cpu_mmu_enable_icache()` also.
We may need to review all the other places where we invalidate the TLB, but that out of scope for this PR at least.

Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: ok

**Comment by etienne-lms** [comment]
> When CFG_SCTLR_RR and CFG_SCTLR_Z are y, the bits are set ; but they are not cleared when the config is n (contrary to what happens with CFG_SCTLR_A). What is the correct way?


Maybe we could change `CFG_SCTLR_A|RR|A` to 0|1|undefined:
- 0 => clear SCTLR[x] bit
- 1 => clear SCTLR[x]  set bit
- undefined (=='n') =>don't modify SCTLR[x] bit

**Comment by jenswi-linaro** [comment]
Contrary to the A bit I don't see any reason to clear the RR and Z bits. To make it more clear we could squeeze in an `_ENABLE_` in the config flag.

**Comment by jforissier** [comment]
@jenswi-linaro sounds good to me, so, @etienne-lms how about just replacing:
```
s/CFG_SCTLR_ROUNDROBIN_REPLACEMENT/CFG_ENABLE_SCTLR_RR/
s/CFG_EARLY_BRANCH_PREDICTION/CFG_ENABLE_SCTLR_Z/
```
?

**Comment by etienne-lms** [comment]
I am confused but some reported issues since the sequence is no more the same for the modified platforms. Alignment is now effective (which is good) and some SCTLR bit field may have a modified value since the SCTLR is no more reset to a given value from the several `plat_cpu_reset_early()`.

I would propose another approach, restore the code i've removed from some platforms and call `plat_cpu_reset_early()` prior to SCTLR config from generic_entry_a32.S. Such change would insure this later change in 2.5.0 only enables SCTLR[A], versus 2.5.0-rc1.

Another option is to postpone this change to after the 2.5.0 tag :(

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 2]
```diff
@@ -473,8 +484,17 @@ UNWIND(	.cantunwind)
 #if defined(CFG_HWSUPP_MEM_PERM_WXN) && defined(CFG_CORE_RWDATA_NOEXEC)
 	orr	r0, r0, #(SCTLR_WXN | SCTLR_UWXN)
 #endif
+#if defined(CFG_SCTLR_ENABLE_RR)
+	orr	r0, r0, #SCTLR_RR
+#endif
 	write_sctlr r0
 
+#if defined(CFG_SCTLR_ENABLE_Z)
+	write_bpiall
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: Need same
```
dsb
isb
```
sequence here

**Comment by etienne-lms** [comment]
This P-R is related to discussions around https://github.com/OP-TEE/optee_os/pull/1638.
Thanks, sharing gave ideas, but there are potential regressions that can't match the initial target of 2.5.0.

I will close this P-R,  come back to the discussion from https://github.com/OP-TEE/optee_os/pull/1638 and propose other P-R if required.

**Pull request closed without merging**